presc/training_stability/cross_validation.py

```diff
@@ -0,0 +1,121 @@
+    for small numbers of k-folds is to run the cross validation several times
+    and average all those values.
+
+    Parameters:
+        dataset (DataFrame): Pandas dataset.
+        pipeline (Pipeline): Can be a scaler and a classifier or only the
+            classifier.
+        metrics (list of str): So far only "accuracy" is possible, although the
+            fit and scoring times are also computed.
+        kfolds_list (list): List of the number of k-folds for which to compute
+            the cross-validation scores.
+        repetitions (int): Number of times to repeat the cross-validation for
+            each k-fold number. By default it carries out the cross-validation
+            once for each requested number of k-folds.
+        minimum_repetitions (bool): If "minimum_repetitions" is set to True,
+            "repetitions" becomes the minimum number of repetitions and the
+            number of needed cross-validation runs for each k-fold value are
+            carried out. For example, the cross validation for five k-folds
+            yields five performance values so, if the minimum number of
+            repetitions is set to ten, the cross validation will be repeated
+            twice.
+    Returns:
+        kfolds_list (list): List of the number of k-folds for which to compute
+            the cross-validation scores.
+        scores_summary (dict of list): Dictionary with the different computed
+            scores. Each score has a list with the list of average values for
+            each k-fold value, and the list of standard deaviations for each
+            k-fold value.
+    """
+
+    X = dataset.iloc[:, :-1]
+    y = dataset.iloc[:, -1]
+
+    repetitions_original = repetitions
+
+    scores_summary = None
+
+    for num in kfolds_list:
+
+        # If number of k-folds is smaller than the minimum number of
+        # repetitions, repeat the cross-validation for that k-fold number as
+        # many times as necessary.
+        if minimum_repetitions is True:
+            if num <= repetitions_original:
+                repetitions = ceil(repetitions_original / num)
+            else:
+                repetitions = 1
+
+        scores_list = []
+        scores_kfold = []
+
+        for repetition in range(repetitions):
+            cv = KFold(n_splits=num, shuffle=True)  # , random_state=None)
+            scores = cross_validate(
+                pipeline, X, y, scoring=metrics, cv=cv, return_estimator=False
+            )
+            scores_list += [scores]
+
+        # Gather all repetitions of each kfold
+        scores_kfold = {metric: [] for metric in scores}
+
+        for element in scores_list:
+            for metric in element:
+                scores_kfold[metric] += list(element[metric])
+
+        # Consolidate all repetitions
+        scores_kfold = {
+            metric: [np.mean(scores_kfold[metric]), np.std(scores_kfold[metric])]
```

Comment:
If we were running CV without repeating, `scores` gives you the score values within each fold, and then we would average. The goal here is to assess how much that average can vary across the repetitions. To do that, we would need to compute the average score for each rep and then find the SD of those averages, rather than the SD across all indidivual fold scores. The difference being that the scores across folds within one rep are not independent.

Consequently, running this function with 1 rep would give us the trend across different fold numbers but no estimate of variability. I recommend setting the default rep value to at least 10.